File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.ts
function distinctSubseqII(s: string): number {
    const MOD = 1_000_000_007;
    const endWith = new Array(26).fill(0);
    let total = 0;

    for (const ch of s) {
        const idx = ch.charCodeAt(0) - 97;
        const newSubseqs = (1 + total - endWith[idx] + MOD) % MOD;
        total = (total + newSubseqs) % MOD;
        endWith[idx] = (endWith[idx] + newSubseqs) % MOD;
    }
    return total;
};

File: 0940-distinct-subsequences-ii/0940-distinct-subsequences-ii.cpp
class Solution {
public:
    int distinctSubseqII(string s) {
        const int MOD = 1e9 + 7;
        vector<int> countEndWith(26, 0);
        int total = 0;

        for (char c : s) {
            int idx = c - 'a';
            int newSubseqs = (1 + total - countEndWith[idx] + MOD) % MOD;
            total = (total + newSubseqs) % MOD;
            countEndWith[idx] = (countEndWith[idx] + newSubseqs) % MOD;
        }
        return total;
    }
};